### PR TITLE
fix: Compilation fails if Default*Pairs unimplemented in service

### PR DESCRIPTION
**Pull request by JinnyYi**
fix https://github.com/beyondstorage/go-storage/issues/760

### CODE REVIEW

**Comment by Xuanwo** [comment]
Maybe `default_storage_pairs` and `storage_features` should be required?

How about always add `default_storage_pairs` and `storage_features` for services?

**Comment by JinnyYi** [comment]
> Maybe `default_storage_pairs` and `storage_features` should be required?
> 
> How about always add `default_storage_pairs` and `storage_features` for services?

Does it mean `default_storage_pairs` should be added in `namespace.storage.new`  for all services? Or should we generate `DefaultStoragePairs` in `pairStorageNew`  struct no matter there's defalt pair or not?

For `storage_features`, no additional logic is currently required to handle it.

**Comment by Xuanwo** [comment]
> Does it mean `default_storage_pairs` should be added in `namespace.storage.new` for all services?

Yes, I think this way could be better.

**Comment by JinnyYi** [comment]
> > Does it mean `default_storage_pairs` should be added in `namespace.storage.new` for all services?
> 
> Yes, I think this way could be better.

Then maybe we don't need to modify `go-storage` for `default_storage_pairs`.

**Comment by Xuanwo** [comment]
> Then maybe we don't need to modify `go-storage` for `default_storage_pairs`.

How about adding them by default? And finally, we can remove `default_storage_pairs` in `services.toml` as they will be auto-added.

File: tests/generated.go [hunk 1]
```diff
@@ -163,14 +163,8 @@ type pairServiceNew struct {
 	HasCredential bool
 	Credential    string
 	// Optional pairs
-	HasDefaultContentType  bool
```
Comment (Xuanwo) [inline tests/generated.go:166]: Why they are been removed?

Comment (JinnyYi) [reply tests/generated.go:166]: There's no operation in `service` namespace has the `content_type` pair. So I modified the way to get the defaultable pairs in https://github.com/beyondstorage/go-storage/pull/821/files#diff-0ca1d3b96d3cc7d55c46d876668a196e55b5a73e0208c01307c5d93e93d0bc9cR311-R312

Comment (Xuanwo) [reply tests/generated.go:166]: Makes sense.

File: cmd/definitions/type.go [hunk 2]
```diff
@@ -307,14 +308,34 @@ func (d *Data) LoadService(filePath string) {
...
+			ns.New.Optional = append(ns.New.Optional, "default_"+v.Pair.Name)
+		}
+		// add default_*_pairs and *_features
+		hasDefaultNsPairs := false
+		hasNsFeatures := false
+		defaultNsPairsName := fmt.Sprintf("default_%s_pairs", ns.Name)
+		nsFeaturesName := fmt.Sprintf("%s_features", ns.Name)
+		for _, v := range ns.New.Optional {
+			if v == defaultNsPairsName {
+				hasDefaultNsPairs = true
```
Comment (Xuanwo) [inline cmd/definitions/type.go:321]: How about adding a warning here to let maintain remove the `default_storage_pairs` and `storage_features`? As we will auto generate for them.

**Review by Xuanwo — APPROVED** [review]

File: cmd/definitions/type.go (hunk 2 above)
Comment (JinnyYi) [reply cmd/definitions/type.go:322]: Maybe I also should remove the pairs from tests.

**Pull request merged**